File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
        HashMap<Character,Integer>mp = new HashMap<>();
        int n=s.length();
        for(int i=0;i<n;i++){
            char ch=s.charAt(i);
            mp.put(ch,mp.getOrDefault(ch,0)+1);
        }
        ArrayList<Map.Entry<Character,Integer>>result = new ArrayList<>(mp.entrySet());
        result.sort((a,b)->{
            if(a.getValue()==b.getValue()){
                return a.getKey()-b.getKey();
            }
            return b.getValue()-a.getValue();
        });

      
        String ans="";
        for(Map.Entry<Character,Integer>entry: result){
            for(int i=0;i<entry.getValue();i++){
                ans+=entry.getKey();
            }
        }
        return ans;
    }
}

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        int n=s.size();
        unordered_map<char,int>mp;
        for(int i=0;i<n;i++){
            mp[s[i]]++;
        }
        vector<pair<char,int>>list(mp.begin(),mp.end());
        sort(list.begin(),list.end(),[](auto &a, auto &b){
            if(a.second==b.second){
                return a.first< b.first;
            }
            return a.second>b.second;
        });
        string ans="";
        for(auto &it: list){
            for(int i=0;i<it.second;i++){
                ans+=it.first;
            }
        }
        return ans;
    }
};